# Aufgabe 1

Eine Boeing 737-200 setzt bei der Landung zum Zeitpunkt $t = O \text{s}$ bei der Koordinate $x_0 = 0 \text{m}$ mit der Geschwindigkeit $v_0 = 100 \frac{m}{s}$ auf. Die Masse des Flugzeugs beträgt zum Zeitpunkt der Landung $m = 97'000 \text{kg}$ und wird im Weiteren als konstant angenommen. Die durch Schubumkehr erzeugte Bremskraft ist gegeben durch
$F —5 \cdot \dot{x}^2 — 570000$ und daraus folgt als Bewegungsgleichung die DGL 2. Ordnung:

$$
m \cdot \ddot{x} = -5 \cdot \dot{x}^2 - 570000
$$

wobei $x(t)$ die Ortsfunktion des Flugzeuges als Funktion der Zeit $t$ ist, $\dot{x} = \frac{dx}{dt} = v(t)$ seine Geschwindigkeit und $\ddot{x} = \frac{d^2x}{dt^2} = \frac{dv}{dt}= a(t)$ die Beschleunigung. Damit kann die DGL 2. Ordnung umgeschrieben werden in eine DGL 1. Ordnung:

$$
m \frac{dv}{dt} = -5v^2 - 570000
$$

#### a)

Gesucht ist die Zeit $t_E$, die das Flugzeug braucht, um zum Stillstand zu kommen. Sie werden in den späteren
Übungen Verfahren anwenden, um $v(t)$ und $x(t)$ direkt numerisch berechnen zu können. Bis es soweit ist, benutzen
Sie hier das analytische Verfahren der Separation der Variablen, um ein Integral für die Zeit zu erhalten:

$$
t_E = \int_{0}^{t_E} \ dt = ?
$$

Berechnen Sie das erhaltene Integral numerisch mit einem Skript `Name_S10_Aufg2.py` mit der Romberg-Extrapolation aus Serie 9, Aufgabe 3).

In [14]:
import numpy as np

def WIN07_S9_Aufg3(f, a, b, m):
    """
    a: Untere Grenze Intervall (float / int)
    b: Obere Grenze Intervall (float / int)
    f: Funktion für welche das Integral abgeschätzt werden soll (lambda oder Funktion)
    m: Anzahl Stützstellen (int)
    """
    first_row = np.array([])

    for j in range(m):
        h = (b - a) / 2**j # Intervallbreite
        x_values = np.arange(a, b+h, h)
        sum_trapez = np.array([(f(x)) for _, x in enumerate(x_values[1: -1])]).sum()
        frac_trapez = (f(a) + f(b))/2
        integral_trapez = (h * (frac_trapez + sum_trapez))
        first_row = np.append(first_row, integral_trapez)

    T = np.zeros((m,m))
    T[: , 0] = first_row

    last_row = np.copy(first_row)
    for row in range(1, T.shape[0]):
        next_row = np.array([(4**row * last_row[idx] - last_row[idx-1])/(4**row-1) for idx, _ in enumerate(last_row) if idx > 0])
        last_row = np.copy(next_row)
        for _ in range(T.shape[0] - len(next_row)):
            next_row = np.append(next_row, 0)
        T[:, row] = next_row

    return T[0][-1]

Die Ausgangslage für die Berechnung bildet die folgende Gleichung

$$
m \cdot \ddot{x} = -5 \cdot \dot{x}^2 - 570'000
$$

Nun kann die zweite Ableitung des Ortes $\ddot{x}$ als einfache Ableitung der Geschwindigkeit aufgefasst werden, ausserdem wird die einfache Ableitung des Ortes $\dot{x}$ als Geschwindigkeit $v$ geschrieben:

$$
m \cdot \dot{v} = -5 \cdot v^2 - 570'000
$$

Wir schreiben die Formel in einer verständlicheren Notation:

$$
m \cdot \frac{dv}{dt} = -5 \cdot v^2 - 570'000
$$

Nun müssen alle Terme, die die Geschwindigkeit beinhalten auf eine Seite gebracht werden. Dies machen wir in zwei Schritten:

$$
m \cdot dv = (-5 \cdot v^2 - 570'000) \ dt
$$

$$
\rightarrow \frac{m \cdot dv}{(-5 \cdot v^2 - 570'000)} = dt
$$

Nun kann man ein bestimmtes Integral mit den Grenzwerten $v_0$ und $v_{E} = 0 \frac{m}{s}$ bilden. Dies entspricht dann dem Integral der Zeit von $t_0$ bis $t_E$ :

$$
\int_{100}^{0} \frac{m}{(-5 \cdot v^2 - 570'000)} \ dv = \int_{0}^{t_E}\ dt 
$$

Da die Masse eine Konstante ist, kann sie vor das Integral genommen werden:

$$
m \cdot \int_{100}^{0} \frac{1}{(-5 \cdot v^2 - 570'000)} \ dv = \int_{0}^{t_E}\ dt 
$$

In [19]:
from IPython.display import Markdown as md
m = 97000 # Masse des Flugzeuges
f_v = lambda v: 1 / (-5 * v**2 - 570000) # Funktion, welche abgeleitet werden soll
v0 = 100 #Anfangswert bestimmtes Integral, entspricht b in Romberg-Extrapolation
v_end = 0 #Endwert bestimmtes Integral, entspricht a in Romberg-Extrapolation

#Gemäss obiger Erklärung muss die Masse mit dem berechneten Integral multipliziert werden
approx_t = m * WIN07_S9_Aufg3(f_v, v0, v_end, 10)

display(md(f"##### Approximierte Zeit bis zum Stillstand: ${approx_t}$"))

##### Approximierte Zeit bis zum Stillstand: $16.544607316744216$

Gemäss der obigen Berechnung beträgt die approximierte Zeit, die das Flugzeug bis zum Stillstand benötigt $16.5446\dots \text{m}$

#### b) 

Benutzen Sie den Zusammenhang $\frac{dx}{dt} = v$ resp. $dx = dt \cdot v$ um aus a) analytisch auch das Integral für den Bremsweg $x_E$ zu erhalten

$$
x_E = \int_{0}^{x_E} \ dx = ?
$$

und berechnen Sie auch dieses Integral mit der Romberg-Extrapolation

Wir benutzen den wie oben vorgegebenen Ansatz $dx = dt \cdot v$ und definieren eine neue Lambda-Funktion. Die Grenzwerte des Integrals können wir gleich belassen. Dies ergibt also: 

$$
x_E = \int_{0}^{x_E} \ dx = m \cdot \int_{100}^{0} \frac{1}{(-5 \cdot v^2 - 570000)} \cdot v \ dv
$$

In [16]:
f_h = lambda v: f_v(v) * v
approx_xE = m * WIN07_S9_Aufg3(f_h, v0, v_end, 10)
print(f"Die approximierte zurückgelegte Strecke beträgt {approx_xE} m")

Die approximierte zurückgelegte Strecke beträgt 815.6062369422517 m


Die approximierte Strecke, die das Flugzeug bis zum Stillstand zurücklegt, beträgt gemäss obiger Berechnung ca. $815.60623\dots \text{m}$ 